subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/DefaultMemoryManager.java

```diff
@@ -117,7 +117,7 @@ private long doRequestFreeMemory(long requestedFreeMemory, long freeMemory) {
         if (freeMemory < requestedFreeMemory) {
             LOGGER.debug("{} memory requested, {} free", requestedFreeMemory, freeMemory);
             synchronized (holdersLock) {
-                for (MemoryHolder holder : holders) {
+                for (MemoryHolder holder : new ArrayList<MemoryHolder>(holders)) {
```

Comment:
⭕️  The copy could be done while holding the `holdersLock` and the actual processing out of lock.